File: packages/caelus/src/patterns.ts
/**
 * Classical aspect configurations as first-class objects.
 *
 * Pure geometry over a chart's body longitudes (and houses, for stelliums); no
 * interpretation. Each configuration is judged from pairwise angular separations
 * against an explicit, overridable orb policy: the engine's {@link DEFAULT_ORBS}
 * for the Ptolemaic aspects, plus a quincunx (150°) for yods. The default body
 * set is the aspectable bodies (planets and Chiron; nodes and Lilith excluded).
 *
 * Reported patterns are maximal: a grand cross suppresses the two T-squares it
 * contains and a kite suppresses its grand trine. Port of the Python reference
 * `astroengine.patterns`, pinned by `patterns-golden`.
 */
import { mod } from "./core.js";
import { SIGNS, NOT_ASPECTABLE } from "./chart.js";
import type { Chart } from "./chart.js";

/** Pattern aspects, including the quincunx the default Ptolemaic set omits. */
export const PATTERN_ANGLES: Record<string, number> = {
  conjunction: 0, sextile: 60, square: 90, trine: 120, quincunx: 150, opposition: 180,
};

/** Default orbs: {@link DEFAULT_ORBS} for the five Ptolemaic aspects, plus a
 *  tight quincunx. Override via {@link PatternOptions.orbs}. */
export const PATTERN_ORBS: Record<string, number> = {
  conjunction: 8, sextile: 4, square: 7, trine: 7, quincunx: 3, opposition: 8,
};

const KIND_ORDER = [
  "grand_cross", "mystic_rectangle", "kite", "t_square", "grand_trine",
  "yod", "stellium_sign", "stellium_house",
];

/** One configuration found in a chart. */
export interface ChartPattern {
  /** Configuration kind, e.g. `"t_square"` or `"grand_trine"`. */
  kind: string;
  /** Participating body ids, sorted. */
  bodies: string[];
  /** Focal body for a T-square or yod (the squared / quincunx apex). */
  apex?: string;
  /** Sign for a `stellium_sign`. */
  sign?: string;
  /** House for a `stellium_house`. */
  house?: number;
  /** Worst defining-aspect orb in degrees; `0` for stelliums. */
  orb: number;
}

/** A body's longitude (and house, for stelliums) for {@link detectPatternsIn}. */
export interface PatternBody {
  lon: number;
  house?: number | null;
}

export interface PatternOptions {
  /** Per-aspect orb overrides (degrees), keyed by aspect name. */
  orbs?: Record<string, number>;
  /** Body ids to consider; defaults to the aspectable bodies present. */
  bodies?: string[];
}

const separation = (la: number, lb: number) => Math.abs(mod(la - lb + 180, 360) - 180);

/** The single aspect a pair forms (orbs do not overlap), as `[name, orb]`, or null. */
function relation(la: number, lb: number, orbs: Record<string, number>): [string, number] | null {
  const sep = separation(la, lb);
  for (const name of Object.keys(PATTERN_ANGLES)) {
    const orb = Math.abs(sep - PATTERN_ANGLES[name]);
    if (orb <= orbs[name]) return [name, orb];
  }
  return null;
}

function cmpBodies(a: string[], b: string[]): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) if (a[i] !== b[i]) return a[i] < b[i] ? -1 : 1;
  return a.length - b.length;
}

/**
 * The configurations present among a body map. Lower-level form of
 * {@link detectPatterns}: `bodies` maps a body id to `{ lon, house? }`.
 */
export function detectPatternsIn(
  bodies: Record<string, PatternBody>, opts: PatternOptions = {},
): ChartPattern[] {
  const orbs = opts.orbs ?? PATTERN_ORBS;
  const names = (opts.bodies ?? Object.keys(bodies).filter((b) => !NOT_ASPECTABLE.has(b)))
    .filter((b) => b in bodies);
  const lon: Record<string, number> = {};
  for (const b of names) lon[b] = mod(bodies[b].lon, 360);

  const key = (a: string, b: string) => (a < b ? `${a}|${b}` : `${b}|${a}`);
  const rel = new Map<string, [string, number]>();
  for (let i = 0; i < names.length; i++) {
    for (let j = i + 1; j < names.length; j++) {
      const r = relation(lon[names[i]], lon[names[j]], orbs);
      if (r) rel.set(key(names[i], names[j]), r);
    }
  }
  const asp = (a: string, b: string) => rel.get(key(a, b)) ?? null;
  const isAspect = (a: string, b: string, kind: string) => {
    const r = asp(a, b);
    return r !== null && r[0] === kind;
  };

  const out: ChartPattern[] = [];

  // Grand trines (3-body) and grand crosses / mystic rectangles (4-body).
  const grandTrines: ChartPattern[] = [];
  for (let i = 0; i < names.length; i++) {
    for (let j = i + 1; j < names.length; j++) {
      for (let k = j + 1; k < names.length; k++) {
        const [a, b, c] = [names[i], names[j], names[k]];
        if (isAspect(a, b, "trine") && isAspect(b, c, "trine") && isAspect(a, c, "trine")) {
          const orb = Math.max(asp(a, b)![1], asp(b, c)![1], asp(a, c)![1]);
          grandTrines.push({ kind: "grand_trine", bodies: [a, b, c].sort(), orb });
        }
      }
    }
  }

  const grandCrosses: ChartPattern[] = [];
  const mysticRectangles: ChartPattern[] = [];
  for (let i = 0; i < names.length; i++) {
    for (let j = i + 1; j < names.length; j++) {
      for (let k = j + 1; k < names.length; k++) {
        for (let l = k + 1; l < names.length; l++) {
          const quad = [names[i], names[j], names[k], names[l]];
          const pairs: Array<[string, string]> = [
            [quad[0], quad[1]], [quad[0], quad[2]], [quad[0], quad[3]],
            [quad[1], quad[2]], [quad[1], quad[3]], [quad[2], quad[3]],
          ];
          const kinds = pairs.map(([a, b]) => asp(a, b));
          if (kinds.some((r) => r === null)) continue;
          const counts: Record<string, number> = {};
          let worst = 0;
          for (const r of kinds) {
            counts[r![0]] = (counts[r![0]] ?? 0) + 1;
            if (r![1] > worst) worst = r![1];
          }
          if (counts.opposition === 2 && counts.square === 4) {
            grandCrosses.push({ kind: "grand_cross", bodies: [...quad].sort(), orb: worst });
          } else if (counts.opposition === 2 && counts.trine === 2 && counts.sextile === 2) {
            mysticRectangles.push({ kind: "mystic_rectangle", bodies: [...quad].sort(), orb: worst });
          }
        }
      }
    }
  }

  // Kite: a grand trine plus a fourth body opposite one member.
  const kites: ChartPattern[] = [];
  for (const gt of grandTrines) {
    const tri = gt.bodies;
    for (const d of names) {
      if (tri.includes(d)) continue;
      for (const apex of tri) {
        const others = tri.filter((x) => x !== apex);
        if (isAspect(d, apex, "opposition")
          && isAspect(d, others[0], "sextile")
          && isAspect(d, others[1], "sextile")) {
          const orb = Math.max(gt.orb, asp(d, apex)![1], asp(d, others[0])![1], asp(d, others[1])![1]);
          kites.push({ kind: "kite", bodies: [...tri, d].sort(), apex, orb });
        }
      }
    }
  }

  // T-square: an opposition whose two ends both square a common apex.
  const tSquares: ChartPattern[] = [];
  for (let i = 0; i < names.length; i++) {
    for (let j = i + 1; j < names.length; j++) {
      const [a, b] = [names[i], names[j]];
      if (!isAspect(a, b, "opposition")) continue;
      for (const apex of names) {
        if (apex === a || apex === b) continue;
        if (isAspect(apex, a, "square") && isAspect(apex, b, "square")) {
          const orb = Math.max(asp(a, b)![1], asp(apex, a)![1], asp(apex, b)![1]);
          tSquares.push({ kind: "t_square", bodies: [a, b, apex].sort(), apex, orb });
        }
      }
    }
  }

  // Yod: a sextile whose two ends both quincunx a common apex.
  const yods: ChartPattern[] = [];
  for (let i = 0; i < names.length; i++) {
    for (let j = i + 1; j < names.length; j++) {
      const [a, b] = [names[i], names[j]];
      if (!isAspect(a, b, "sextile")) continue;
      for (const apex of names) {
        if (apex === a || apex === b) continue;
        if (isAspect(apex, a, "quincunx") && isAspect(apex, b, "quincunx")) {
          const orb = Math.max(asp(a, b)![1], asp(apex, a)![1], asp(apex, b)![1]);
          yods.push({ kind: "yod", bodies: [a, b, apex].sort(), apex, orb });
        }
      }
    }
  }

  // Suppress sub-patterns contained in a larger reported one.
  const subset = (small: string[], big: string[]) => small.every((x) => big.includes(x));
  const keptTSquares = tSquares.filter((t) => !grandCrosses.some((g) => subset(t.bodies, g.bodies)));
  const keptTrines = grandTrines.filter((g) => !kites.some((k) => subset(g.bodies, k.bodies)));

  out.push(...grandCrosses, ...mysticRectangles, ...kites, ...keptTSquares, ...keptTrines, ...yods);

  // Stelliums by sign and by house: three or more bodies sharing one.
  const bySign: Record<number, string[]> = {};
  for (const b of names) {
    const s = Math.floor(lon[b] / 30) % 12;
    (bySign[s] ??= []).push(b);
  }
  for (const s of Object.keys(bySign)) {
    const members = bySign[Number(s)];
    if (members.length >= 3) {
      out.push({ kind: "stellium_sign", bodies: [...members].sort(), sign: SIGNS[Number(s)], orb: 0 });
    }
  }

  const byHouse: Record<number, string[]> = {};
  for (const b of names) {
    const h = bodies[b].house;
    if (h != null) (byHouse[h] ??= []).push(b);
  }
  for (const h of Object.keys(byHouse)) {
    const members = byHouse[Number(h)];
    if (members.length >= 3) {
      out.push({ kind: "stellium_house", bodies: [...members].sort(), house: Number(h), orb: 0 });
    }
  }

  out.sort((a, b) => KIND_ORDER.indexOf(a.kind) - KIND_ORDER.indexOf(b.kind) || cmpBodies(a.bodies, b.bodies));
  for (const p of out) p.orb = Math.round(p.orb * 1e4) / 1e4;
  return out;
}

/**
 * The aspect configurations present in a {@link Chart}: T-squares, grand trines,
 * grand crosses, yods, kites, mystic rectangles, and stelliums by sign and by
 * house. Bodies outside their fitted range (absent from the chart) are skipped.
 *
 * @param chart A {@link Chart} from {@link Engine.chart} / {@link Engine.chartAt}.
 * @param opts Orb overrides and an optional explicit body set.
 * @returns The configurations, most-complex first, each a {@link ChartPattern}.
 * @example
 * ```ts
 * const chart = engine.chart(1990, 6, 10, 14, 30, 0, 27.95, -82.46, "placidus");
 * detectPatterns(chart); // [{ kind: "mystic_rectangle", bodies: [...], orb: 2.54 }, ...]
 * ```
 */
export function detectPatterns(chart: Chart, opts: PatternOptions = {}): ChartPattern[] {
  const bodies: Record<string, PatternBody> = {};
  for (const [name, p] of Object.entries(chart.bodies)) {
    if (p) bodies[name] = { lon: p.lon, house: p.house };
  }
  return detectPatternsIn(bodies, opts);
}

File: python/astroengine/patterns.py
"""astroengine.patterns -- classical aspect configurations as first-class objects.

Pure geometry over a chart's body longitudes (and houses, for stelliums): no
interpretation. Each configuration is judged from pairwise angular separations
against an explicit, overridable orb policy (the engine's DEFAULT_ORBS for the
Ptolemaic aspects, plus a quincunx for yods). The default body set is the
aspectable bodies (planets and Chiron; nodes and Lilith are excluded, matching
find_aspects).

Reported patterns are maximal: a grand cross suppresses the two T-squares it
contains, and a kite suppresses its grand trine, so the output reads the way an
astrologer would name it. The TS port (patterns.ts) reproduces every value and
the golden fixtures pin the two together.
"""
from itertools import combinations

from .chart import NOT_ASPECTABLE, SIGNS

# Pattern aspects, including the quincunx (150 deg) that yods need and that the
# default Ptolemaic set omits.
PATTERN_ANGLES = {
    "conjunction": 0.0, "sextile": 60.0, "square": 90.0,
    "trine": 120.0, "quincunx": 150.0, "opposition": 180.0,
}
# Default orbs: the engine's DEFAULT_ORBS for the five Ptolemaic aspects, plus a
# tight quincunx. Overridable via the `orbs` argument.
PATTERN_ORBS = {
    "conjunction": 8.0, "sextile": 4.0, "square": 7.0,
    "trine": 7.0, "quincunx": 3.0, "opposition": 8.0,
}
# Stable output order, most-complex first.
_KIND_ORDER = [
    "grand_cross", "mystic_rectangle", "kite", "t_square", "grand_trine",
    "yod", "stellium_sign", "stellium_house",
]


def _separation(la, lb):
    return abs((la - lb + 180.0) % 360.0 - 180.0)


def _relation(la, lb, orbs):
    """The single aspect a pair forms (orbs do not overlap), as (name, orb), or
    None. Conjunction is tracked but forms no multi-body pattern here."""
    sep = _separation(la, lb)
    for name, angle in PATTERN_ANGLES.items():
        orb = abs(sep - angle)
        if orb <= orbs[name]:
            return (name, orb)
    return None


def detect_patterns(bodies, orbs=None, body_filter=None):
    """Configurations present among ``bodies``.

    ``bodies`` maps a body id to ``{"lon": deg, "house": int|None}``. ``orbs``
    overrides PATTERN_ORBS. ``body_filter`` overrides the default aspectable set.
    Returns a list of pattern dicts in a fixed order; each has ``kind``,
    ``bodies`` (sorted), ``orb`` (worst defining aspect, 0 for stelliums), and
    ``apex``/``sign``/``house`` where they apply.
    """
    orbs = orbs or PATTERN_ORBS
    names = body_filter if body_filter is not None else [
        b for b in bodies if b not in NOT_ASPECTABLE
    ]
    names = [b for b in names if b in bodies]
    lon = {b: bodies[b]["lon"] % 360.0 for b in names}

    # Pairwise aspect relations, computed once.
    rel = {}
    for a, b in combinations(names, 2):
        r = _relation(lon[a], lon[b], orbs)
        if r is not None:
            rel[frozenset((a, b))] = r

    def asp(a, b):
        return rel.get(frozenset((a, b)))

    def is_aspect(a, b, kind):
        r = asp(a, b)
        return r is not None and r[0] == kind

    out = []

    # Grand trines and mystic rectangles / grand crosses come from 3- and
    # 4-body combinations; T-squares and yods from an opposition/sextile base
    # plus an apex.
    grand_trines = []
    for tri in combinations(names, 3):
        a, b, c = tri
        if is_aspect(a, b, "trine") and is_aspect(b, c, "trine") and is_aspect(a, c, "trine"):
            orb = max(asp(a, b)[1], asp(b, c)[1], asp(a, c)[1])
            grand_trines.append({"kind": "grand_trine", "bodies": sorted(tri), "orb": orb})

    grand_crosses = []
    mystic_rectangles = []
    kites = []
    for quad in combinations(names, 4):
        pairs = list(combinations(quad, 2))
        kinds = [asp(*p) for p in pairs]
        if any(k is None for k in kinds):
            continue
        counts = {}
        for k in kinds:
            counts[k[0]] = counts.get(k[0], 0) + 1
        worst = max(k[1] for k in kinds)
        if counts.get("opposition") == 2 and counts.get("square") == 4:
            grand_crosses.append({"kind": "grand_cross", "bodies": sorted(quad), "orb": worst})
        elif counts.get("opposition") == 2 and counts.get("trine") == 2 and counts.get("sextile") == 2:
            mystic_rectangles.append({"kind": "mystic_rectangle", "bodies": sorted(quad), "orb": worst})

    # Kite: a grand trine plus a fourth body opposite one member (and so sextile
    # the other two).
    for gt in grand_trines:
        tri = gt["bodies"]
        for d in names:
            if d in tri:
                continue
            for apex in tri:
                others = [x for x in tri if x != apex]
                if (is_aspect(d, apex, "opposition")
                        and is_aspect(d, others[0], "sextile")
                        and is_aspect(d, others[1], "sextile")):
                    orb = max(gt["orb"], asp(d, apex)[1], asp(d, others[0])[1], asp(d, others[1])[1])
                    kites.append({"kind": "kite", "bodies": sorted(tri + [d]),
                                  "apex": apex, "orb": orb})

    # T-square: an opposition whose two ends both square a common apex.
    t_squares = []
    for a, b in combinations(names, 2):
        if not is_aspect(a, b, "opposition"):
            continue
        for apex in names:
            if apex in (a, b):
                continue
            if is_aspect(apex, a, "square") and is_aspect(apex, b, "square"):
                orb = max(asp(a, b)[1], asp(apex, a)[1], asp(apex, b)[1])
                t_squares.append({"kind": "t_square", "bodies": sorted([a, b, apex]),
                                  "apex": apex, "orb": orb})

    # Yod: a sextile whose two ends both quincunx a common apex.
    yods = []
    for a, b in combinations(names, 2):
        if not is_aspect(a, b, "sextile"):
            continue
        for apex in names:
            if apex in (a, b):
                continue
            if is_aspect(apex, a, "quincunx") and is_aspect(apex, b, "quincunx"):
                orb = max(asp(a, b)[1], asp(apex, a)[1], asp(apex, b)[1])
                yods.append({"kind": "yod", "bodies": sorted([a, b, apex]),
                             "apex": apex, "orb": orb})

    # Suppress sub-patterns contained in a larger reported one.
    cross_sets = [frozenset(g["bodies"]) for g in grand_crosses]
    kite_sets = [frozenset(k["bodies"]) for k in kites]
    t_squares = [t for t in t_squares
                 if not any(frozenset(t["bodies"]) <= cs for cs in cross_sets)]
    grand_trines = [g for g in grand_trines
                    if not any(frozenset(g["bodies"]) <= ks for ks in kite_sets)]

    out += grand_crosses + mystic_rectangles + kites + t_squares + grand_trines + yods

    # Stelliums by sign and by house: three or more bodies sharing one.
    by_sign = {}
    for b in names:
        s = int(lon[b] // 30) % 12
        by_sign.setdefault(s, []).append(b)
    for s, members in by_sign.items():
        if len(members) >= 3:
            out.append({"kind": "stellium_sign", "bodies": sorted(members),
                        "sign": SIGNS[s], "orb": 0.0})

    by_house = {}
    for b in names:
        h = bodies[b].get("house")
        if h is not None:
            by_house.setdefault(h, []).append(b)
    for h, members in by_house.items():
        if len(members) >= 3:
            out.append({"kind": "stellium_house", "bodies": sorted(members),
                        "house": h, "orb": 0.0})

    out.sort(key=lambda p: (_KIND_ORDER.index(p["kind"]), p["bodies"]))
    for p in out:
        p["orb"] = round(p["orb"], 4)
    return out
